### PR TITLE
Pr graduation

### PR DESCRIPTION
**Pull request by kwryankrattiger**
@scottwittenburg

Reviving #45

### CODE REVIEW

File: spackbot/routes.py [hunk 1]
```diff
@@ -125,3 +130,10 @@ async def label_pull_requests(event, gh, *args, session, **kwargs):
     Add labels to PRs based on which files were modified.
     """
     await handlers.add_labels(event, gh)
+
+@router.register("pull_request", action="closed")
+async def on_closed_pull_request(event, gh, *args, session, **kwargs):
+    """
+    Respond to the pull request closed
+    """
+    await handlers.graduate_pr_binaries(event, gh)
```
Comment (scottwittenburg) [inline spackbot/routes.py:139]: I think `handlers/__init__.py` still needs to import this to make it available as you've invoked it here.

Comment (scottwittenburg) [reply spackbot/routes.py:139]: so i.e.:

```
from .mirrors import graduate_pr_binaries
```

**Review by scottwittenburg — CHANGES_REQUESTED** [review]
Thanks @kwryankrattiger, this looks like a lot of work, but is coming along nicely.  I have added what I think are mostly questions, but also some requests.  Hopefully you can straighten me out where I'm not getting it.

Also, a blanket request to apply to the change as a whole:  Can you add docstrings to the new methods and describe the behavior of the methods, as well as expected inputs, and outputs?

Also, before it's all over, you're going to need to run `black` on your changes to get this to pass the style check.  Just install it with `pip` and run `black .` from the top level of the repo, or maybe from the `spackbot` directory, I forget.

File: spackbot/workers.py [hunk 2]
```diff
@@ -328,3 +326,217 @@ async def fix_style_task(event):
...
+            new_obj = shared_pr_bucket.Object(
+                obj.key.replace(pr_mirror_prefix, shared_pr_mirror_prefix, 1)
+            )
+            # Copy the PR mirror object to the new object in the shared PR mirror
+            new_obj.copy(
+                {
+                    "Bucket": pr_bucket_name,
+                    "Key": obj.key,
+                }
+            )
```
[the remark this answers is not in the document]
Comment (kwryankrattiger) [reply spackbot/workers.py:364]: I also dislike this API...quite a shame

File: spackbot/workers.py [hunk 3]
```diff
@@ -328,3 +326,217 @@ async def fix_style_task(event):
...
+                # Prune obj.last_modified > 7 days to avoid storing cached objects
+                # that only existed during development.
+                if (now - last_modified).days >= helpers.pr_mirror_retire_after_days:
+                    logger.debug(
+                        f"pr mirror pruning {obj.key} from s3://{pr_bucket_name}: "
+                        "reason(age)"
+                    )
+                    # Anything older than the retirement age should just be indesciminately
+                    # pruned
+                    obj.delete()
```
[the remark this answers is not in the document]
Comment (scottwittenburg) [reply spackbot/workers.py:446]: Ok, as you pointed out, this is just a naming confusiong on my part.  I saw the `pr_bucket` variable in the loop and assumed it was the mirror for PR, not the shared pr mirror, but I get it now.  I think the argument to the method should be called `shared_pr_mirror_url` and all the `pr_...` things changed to `shared_pr...` for clarity though.

File: spackbot/workers.py [hunk 4]
```diff
@@ -328,3 +326,217 @@ async def fix_style_task(event):
...
+                    # it are also removed.
+                    spec_hash = hash_from_key(obj.key)
+                    if spec_hash:
+                        pr_specs.add(hash_from_key(obj.key))
+                    continue
+
+                if not obj.key.endswith(extensions):
+                    continue
+
+                # Get the hashes in the shared PR bucket.
```
[the remark this answers is not in the document]
Comment (scottwittenburg) [reply spackbot/workers.py:458]: Nvm, this is more naming confusion on my part.

File: spackbot/workers.py [hunk 5]
```diff
@@ -328,3 +326,217 @@ async def fix_style_task(event):
...
+                    logger.error(
+                        f"Encountered spec file without hash in name: {obj.key}"
+                    )
+
+            # Check in the published base branch bucket for duplicates to delete
+            delete_specs = set()
+            for obj in publish_bucket.objects.filter(
+                Prefix=publish_mirror_prefix,
+            ):
+                if not obj.key.endswith(extensions):
```
[the remark this answers is not in the document]
Comment (kwryankrattiger) [reply spackbot/workers.py:472]: Skipping .spack is just to avoid duplicates, but it isn't strictly necessary since it is a set.

**Review by scottwittenburg — APPROVED** [review]
Looks good to me, thanks for getting this done!

**Pull request merged**